File: BinaryTree/sum_tree.py
# sum_tree.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

class Solution:
    def toSumTree(self, root):
        self.convert_to_sum_tree(root)
    
    def convert_to_sum_tree(self, node):
        if not node:
            return 0
        
        # Post-order traversal
        left_sum = self.convert_to_sum_tree(node.left)
        right_sum = self.convert_to_sum_tree(node.right)
        
        # Store original value
        original_value = node.data
        
        # Update node value: 0 for leaf, else sum of subtrees
        node.data = 0 if not node.left and not node.right else left_sum + right_sum
        
        # Return original subtree sum
        return original_value + left_sum + right_sum




#Time Complexity: ( O(N) ), as each node is visited exactly once during post-order traversal.
#Space Complexity: ( O(h) ), where ( h ) is the tree height, due to the recursion stack. For a balanced tree, h=log⁡Nh = \log Nh = \log N; for a skewed tree, h=Nh = Nh = N
.
#Auxiliary Space: ( O(h) ), as no additional data structures are used beyond the recursion stack.
#Best/Average/Worst Case: Time is always ( O(N) ), as every node is processed once. Space is O(log⁡N)O(\log N)O(\log N)for balanced trees and ( O(N) ) for skewed trees.

File: BinaryTree/SumTree.cs
// SumTree.cs
namespace SumTreeSolution
{
    public class Node
    {
        public int Data { get; set; }
        public Node Left { get; set; }
        public Node Right { get; set; }

        public Node(int val)
        {
            Data = val;
            Left = null;
            Right = null;
        }
    }

    public class Solution
    {
        public void ToSumTree(Node root)
        {
            _ = ConvertToSumTree(root);
        }

        private int ConvertToSumTree(Node node)
        {
            if (node == null)
            {
                return 0;
            }

            int leftSum = ConvertToSumTree(node.Left);
            int rightSum = ConvertToSumTree(node.Right);
            int originalValue = node.Data;
            node.Data = (node.Left == null && node.Right == null) ? 0 : leftSum + rightSum;
            return originalValue + leftSum + rightSum;
        }
    }
}
